src/services/accounts/AccountsAssetsService.ts

```diff
@@ -0,0 +1,141 @@
+	 * If none are queried the function will get all AssetId's associated with the
+	 * given `AccountId`, and send back all the `AssetsBalance`s.
+	 *
+	 * @param hash `BlockHash` to make call at
+	 * @param address `AccountId` associated with the balances
+	 * @param assets An array of `assetId`'s to be queried. If the length is zero
+	 * all assetId's associated to the account will be queried
+	 */
+	async fetchAssetBalances(
+		hash: BlockHash,
+		address: string,
+		assets: number[]
+	): Promise<IAccountAssetsBalances> {
+		const { api } = this;
+
+		const { number } = await api.rpc.chain.getHeader(hash);
+
+		let response;
+		if (assets.length === 0) {
+			/**
+			 * This will query all assets and return them in an array
+			 */
+			const keys = await api.query.assets.account.keysAt(hash);
+			const assetIds = this.extractAssetIds(keys);
+
+			response = await this.queryAssets(api, assetIds, address);
+		} else {
+			/**
+			 * This will query all assets by the requested AssetIds
+			 */
+			response = await this.queryAssets(api, assets, address);
+		}
+
+		const at = {
+			hash,
+			height: number.unwrap().toString(10),
+		};
+
+		return {
+			at,
+			assets: response,
+		};
+	}
+
+	/**
+	 * Fetch all `AccountApproval`'s with a given `AssetId` and a `AssetApprovalKey`
+	 * which consists of a `delegate` and an `owner(AccoundId)`
+	 *
+	 * @param hash `BlockHash` to make call at
+	 * @param address `AccountId` or owner associated with the approvals
+	 * @param assetId `AssetId` associated with the `AssetApproval`
+	 * @param delegate `delegate`
+	 */
+	async fetchAssetApproval(
+		hash: BlockHash,
+		address: string,
+		assetId: number,
+		delegate: string
+	): Promise<IAccountAssetApproval> {
+		const { api } = this;
+
+		/**
+		 * AssetApprovalKey, contains the `accountId` as the address key, and the
+		 * delegate `accountId`. Both are required for a successfull query
+		 */
+		const approvalKey = { owner: address, delegate };
+
+		const [{ number }, assetApproval] = await Promise.all([
+			api.rpc.chain.getHeader(hash),
+			api.query.assets.approvals(assetId, approvalKey),
+		]);
+
+		const { amount, deposit } = assetApproval.unwrap();
+
+		const at = {
+			hash,
+			height: number.unwrap().toString(10),
+		};
+
+		return {
+			at,
+			amount,
+			deposit,
+		};
+	}
+
+	/**
+	 * Takes in an array of AssetId's, and an AccountId and returns
+	 * all balances tied to those AssetId's.
+	 *
+	 * @param api ApiPromise
+	 * @param assets An Array of `AssetId`'s or numbers representing assetId's
+	 * @param address An `AccountId` associated with the queried path
+	 */
+	async queryAssets(
+		api: ApiPromise,
+		assets: AssetId[] | number[],
+		address: string
+	): Promise<IAssetBalance[]> {
+		const queryInputAssets = async (): Promise<AssetBalance[]> =>
+			await Promise.all(
+				assets.map((assetId: number | AnyNumber) =>
+					api.query.assets.account(assetId, address)
```

Comment:
Instead of the extra loop below could we make this an `async` call back that like:

```javascript
Promise.all(
   assets.map(async (assetId) => {
      const assetBalance = api.query.assets.account(assetId, address);
     // could we just have `return assetBalance` - unclear to me what the below adds and if it misses `extra`
      return { 
   	            assetId,
		    balance: assetBalance.balance,
		    isFrozen: assetBalance.isFrozen,
		    isSufficient: assetBalance.isSufficient,    
     }
   }

Reply:
If we did it this way, I dont think we would need to wrap it in the `Promise.all`. The purpose for the map here is to create all the calls for the `Promise.all`. My whole idea was trying to avoid making individual queries to the chain, which is why that logic below had to be added as well.

Reply:
It would still need to be wrapped in a `Promise.all` bc the async callback itself returns a `Promise` and map will not wait for the promise to resolve. Query complexity should not increase since they will still be parallel-ized.

Reply:
Awesome, I see what you are saying. Fixed a refactored, thanks.